presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java

```diff
@@ -843,48 +846,71 @@ else if (expression instanceof LongLiteral) {
-                // first, see if this is an ordinal
-                FieldOrExpression groupByExpression;
-
-                if (expression instanceof LongLiteral) {
-                    long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > outputExpressions.size()) {
-                        throw new SemanticException(INVALID_ORDINAL, expression, "GROUP BY position %s is not in select list", ordinal);
-                    }
+        if (node.getGroupBy().isEmpty()) {
+            analysis.setGroupingSets(node, emptyList());
+            return emptyList();
+        }
 
-                    groupByExpression = outputExpressions.get((int) (ordinal - 1));
-                }
-                else {
-                    ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, tupleDescriptor, context);
-                    analysis.addInPredicates(node, expressionAnalysis.getSubqueryInPredicates());
-                    groupByExpression = new FieldOrExpression(expression);
-                }
+        ImmutableList.Builder<List<FieldOrExpression>> groupingSetsBuilder = ImmutableList.builder();
 
-                Type type;
-                if (groupByExpression.isExpression()) {
-                    Analyzer.verifyNoAggregatesOrWindowFunctions(metadata, groupByExpression.getExpression(), "GROUP BY");
-                    type = analysis.getType(groupByExpression.getExpression());
-                }
-                else {
-                    type = tupleDescriptor.getFieldByIndex(groupByExpression.getFieldIndex()).getType();
-                }
-                if (!type.isComparable()) {
-                    throw new SemanticException(TYPE_MISMATCH, node, "%s is not comparable, and therefore cannot be used in GROUP BY", type);
-                }
+        if (node.getGroupBy().size() > 1) {
+            throw new SemanticException(NOT_SUPPORTED, node, "Multiple CUBE/ROLLUP/GROUPING SETs are not yet supported");
+        }
 
-                groupByExpressionsBuilder.add(groupByExpression);
+        for (GroupBySpecification groupBySpecification : node.getGroupBy()) {
+            for (List<Expression> groupingSet : groupBySpecification.enumerateGroupingSets()) {
+                // TODO: handle multiple groupBySpecifications by computing cross-product
+                groupingSetsBuilder.add(analyzeGroupBySpecification(node, tupleDescriptor, context, outputExpressions, groupingSet));
             }
         }
 
-        List<FieldOrExpression> groupByExpressions = groupByExpressionsBuilder.build();
-        analysis.setGroupByExpressions(node, groupByExpressions);
-        return groupByExpressions;
+        List<List<FieldOrExpression>> groupingSets = groupingSetsBuilder.build();
+        if (groupingSets.size() > 1) {
+            throw new SemanticException(NOT_SUPPORTED, node, "Grouping by multiple sets of columns is not yet supported");
```

Comment:
This handles the situation where a single CUBE or ROLLUP, when enumerated, generates grouping sets.